kernel/src/drivers/terminal/mod.rs

```diff
@@ -0,0 +1,316 @@
+
+    /// Writes a colored character to this terminal
+    fn write_colored(&mut self, char: char, color: TerminalColor) -> Result<(), TerminalWriteError<T>>;
+
+    /// Writes a colored string to this terminal
+    fn write_string_colored(&mut self, str: &str, color: TerminalColor) -> Result<(), TerminalWriteError<T>>;
+
+    /// Backspaces the previously written character in this terminal
+    fn backspace(&mut self) -> Result<(), TerminalWriteError<T>>;
+}
+
+/// A writable terminal with bounds and a 2d cursor that can be moved.
+pub trait SizedTerminal<'a, T: TerminalWriter + 'a>: Terminal<'a, T> {
+    /// Sets the position of this terminal's cursor
+    fn set_cursor(&mut self, point: Point) -> Result<(), TerminalWriteError<T>>;
+
+    /// Moves the cursor down on y and returns to origin on x
+    fn new_line(&mut self) -> Result<(), TerminalWriteError<T>>;
+
+    /// Returns true if the given point is inside this terminal
+    fn is_valid(&self, point: Point) -> bool;
+}
+
+/// A raw interface that unsized terminals can write to.
+pub trait TerminalWriter {
+    type Error;
+
+    /// Writes a character through this writer
+    fn write(&mut self, char: TerminalCharacter) -> Result<(), Self::Error>;
+}
+
+/// A raw interface that sized terminals can write to.
+pub trait SizedTerminalWriter: TerminalWriter {
+    /// Sets a character at the given position through this writer
+    fn set(&mut self, point: Point, char: TerminalCharacter) -> Result<(), Self::Error>;
+}
+
+/// Represents a character in a terminal, containing a character and color.
+#[derive(Debug, Copy, Clone)]
+pub struct TerminalCharacter {
+    pub character: char,
+    pub color: TerminalColor,
+}
+
+/// Represents a full color in the terminal with a background and foreground color.
+#[derive(Debug, Copy, Clone)]
+pub struct TerminalColor {
+    pub foreground: Color,
+    pub background: Color,
+}
+
+impl TerminalColor {
+    pub const fn new(foreground: Color, background: Color) -> Self {
+        TerminalColor {
+            foreground: foreground,
+            background: background,
+        }
+    }
+}
+
+/// Represents a 2d point in a terminal.
+#[derive(Debug, Copy, Clone)]
+pub struct Point {
+    pub x: usize,
+    pub y: usize,
+}
+
+impl Point {
+    pub const fn new(x: usize, y: usize) -> Self {
+        Point {
+            x: x,
+            y: y,
+        }
+    }
+}
+
+impl Add<Point> for Point {
+    type Output = Point;
+
+    fn add(self, rhs: Point) -> Point {
+        Point {
+            x: self.x + rhs.x,
+            y: self.y + rhs.y,
+        }
+    }
+}
+
+impl<'a, T: SizedTerminalWriter + 'a> fmt::Write for TextArea<'a, T> {
+    fn write_str(&mut self, str: &str) -> Result<(), fmt::Error> {
+        self.write_string(str).map_err(|_| fmt::Error)?;
+        Ok(())
+    }
+}
+
+/// Writes formatted string to stdout, for macro use
+pub fn stdout_print(args: fmt::Arguments) {
+    use core::fmt::Write;
```

Comment:
This could probably just be a `fmt::Write`... actually, that `use` could be at the top level, since other stuff uses it too.